File: apps/kira-studio/internal/apivars/transforms.go
package apivars

import (
	"encoding/base64"
	"errors"
	"net/url"
	"strings"
	"unicode/utf8"

	"golang.org/x/text/cases"
	"golang.org/x/text/language"
)

// P17 D7/D2: the Go twin of packages/api-core/src/http/transforms.ts — the six-transform closed
// vocabulary a `{{name | transform}}` pipe may apply, byte-identical to the TS side and pinned by
// the same shared corpus (F1). Exactly six, zero-argument, closed (OQ-2 declines arguments).
//
// transformNames is read as plain text by packages/api-core/test/go-ts-api-parity.spec.ts
// (extractGoStringSet) — its shape (a `map[string]bool` literal, one `"name": true` per line) must
// stay exactly this, or that parity check breaks.
var transformNames = map[string]bool{
	"base64":       true,
	"base64decode": true,
	"upper":        true,
	"lower":        true,
	"urlencode":    true,
	"urldecode":    true,
}

// IsTransformName reports whether name is one of the six closed-vocabulary transforms.
func IsTransformName(name string) bool {
	return transformNames[name]
}

// upperCaser/lowerCaser: golang.org/x/text/cases with the "undetermined language" tag applies
// Unicode's *default* case mapping — the same unconditional special-casing rules (ß -> SS, the
// U+FB01 "fi" ligature -> "FI") ECMAScript's String.prototype.toUpperCase/toLowerCase apply,
// unlike strings.ToUpper/ToLower (P21 round 1 finding F4), which only ever does the simple,
// one-rune-to-one-rune Unicode mapping and leaves ß and ligatures untouched.
var (
	upperCaser = cases.Upper(language.Und)
	lowerCaser = cases.Lower(language.Und)
)

// forgivingBase64Decode mirrors the WHATWG "forgiving-base64" algorithm browsers' own atob()
// implements (P21 round 1 finding F4): base64.StdEncoding.DecodeString alone rejects an unpadded
// string ("YQ" for "a") or one with embedded whitespace, both of which atob accepts — so the same
// {{name | base64decode}} pipe transformed a value differently depending on whether the variable
// happened to be marked secret (stage 1, TS/atob) or not (stage 2, Go). ASCII whitespace is
// stripped first, then the remainder is padded out to a multiple of 4 (a length of 1 mod 4 has no
// valid padding and is rejected, matching the WHATWG algorithm exactly).
func forgivingBase64Decode(s string) ([]byte, error) {
	var b strings.Builder
	b.Grow(len(s))
	for _, r := range s {
		switch r {
		case ' ', '\t', '\n', '\f', '\r':
			continue
		}
		b.WriteRune(r)
	}
	stripped := b.String()
	switch len(stripped) % 4 {
	case 1:
		return nil, errors.New("apivars: invalid base64 length")
	case 2:
		stripped += "=="
	case 3:
		stripped += "="
	}
	return base64.StdEncoding.DecodeString(stripped)
}

// applyTransform is one step of ApplyPipeline. D7's table: base64/base64decode round-trip
// standard, padded base64; base64decode additionally requires the decoded bytes be valid UTF-8
// (matching TS's `TextDecoder('utf-8', {fatal: true})`) — invalid base64 *or* invalid UTF-8 is a
// failure (D5). upper/lower never fail. urlencode/urldecode are url.QueryEscape/QueryUnescape
// verbatim — the escaper this repo already uses for a urlencoded body (P12 D7) — except urldecode
// additionally requires the decoded bytes be valid UTF-8 (F4: url.QueryUnescape alone happily
// returns invalid UTF-8 for e.g. `%FF`, where TS's decodeURIComponent throws).
func applyTransform(name, s string) (string, bool) {
	switch name {
	case "base64":
		return base64.StdEncoding.EncodeToString([]byte(s)), true
	case "base64decode":
		decoded, err := forgivingBase64Decode(s)
		if err != nil || !utf8.Valid(decoded) {
			return "", false
		}
		return string(decoded), true
	case "upper":
		return upperCaser.String(s), true
	case "lower":
		return lowerCaser.String(s), true
	case "urlencode":
		return url.QueryEscape(s), true
	case "urldecode":
		decoded, err := url.QueryUnescape(s)
		if err != nil || !utf8.ValidString(decoded) {
			return "", false
		}
		return decoded, true
	default:
		return "", false
	}
}

// ApplyPipeline applies each transform in pipeline to value, left to right (D7's chaining rule),
// returning ok=false the instant any step fails (D5: nothing is emitted half-transformed). An
// empty pipeline returns value unchanged. Every name in pipeline is expected to already be a valid
// transform (ParseReference only ever produces one) — an unrecognised name here is itself treated
// as a failure rather than a panic, so a future caller passing an unvalidated pipeline fails safe.
func ApplyPipeline(pipeline []string, value string) (string, bool) {
	out := value
	for _, name := range pipeline {
		applied, ok := applyTransform(name, out)
		if !ok {
			return "", false
		}
		out = applied
	}
	return out, true
}

File: packages/api-core/src/http/transforms.ts
// P17 D7/D2: the six-transform closed vocabulary a `{{name | transform}}` pipe may apply — zero
// arguments, byte-identical to internal/apivars/transforms.go (the Go twin every transform is
// pinned against by the shared corpus, F1/D2). Pure and dependency-free beyond `./escape` (which
// is itself pure), matching substitute.ts's own stated reason for being a plain, unit-testable
// import.
//
// Exactly six, closed: the SPEC row names base64 encode/decode, upper/lower and URL encode/decode
// and nothing else. A closed vocabulary is what makes D3 rule 4 safe (an unknown `|`-segment must
// be recognisably not a transform, so parseReference can fall back to treating the whole span as a
// literal name) and what keeps Go/TS parity a finite, testable claim (OQ-2 declines arguments).
import { goQueryEscapeLiteral } from './escape';

export const TRANSFORM_NAMES = [
  'base64',
  'base64decode',
  'upper',
  'lower',
  'urlencode',
  'urldecode',
] as const;

export type TransformName = (typeof TRANSFORM_NAMES)[number];

const TRANSFORM_NAME_SET: ReadonlySet<string> = new Set(TRANSFORM_NAMES);

export function isTransformName(name: string): name is TransformName {
  return TRANSFORM_NAME_SET.has(name);
}

// D7: NOT bare `btoa(s)` — btoa throws on any code point > 255 (it operates on a "binary string",
// one byte per char code), so a non-ASCII value would throw in TS where Go's
// base64.StdEncoding.EncodeToString([]byte(s)) happily encodes its UTF-8 bytes. Converting through
// TextEncoder first is what makes the two agree on every input, not just ASCII ones.
function base64Encode(s: string): string | null {
  try {
    const bytes = new TextEncoder().encode(s);
    let binary = '';
    for (const b of bytes) binary += String.fromCharCode(b);
    return btoa(binary);
  } catch {
    return null;
  }
}

// D7: invalid base64 *or* invalid UTF-8 is a failure (D5) — `TextDecoder('utf-8', {fatal: true})`
// is what makes that agree with Go's `utf8.Valid` check; the default (non-fatal) decoder would
// silently substitute U+FFFD for a malformed byte sequence instead of failing, which Go does not.
function base64Decode(s: string): string | null {
  try {
    const binary = atob(s);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
    return new TextDecoder('utf-8', { fatal: true }).decode(bytes);
  } catch {
    return null;
  }
}

// D7: the inverse of goQueryEscapeLiteral — '+' back to a space first (QueryEscape's own space
// handling), then decodeURIComponent for the rest. A malformed `%` escape throws a URIError,
// matching url.QueryUnescape's own error (D5's failure rule).
function urlDecode(s: string): string | null {
  try {
    return decodeURIComponent(s.replace(/\+/g, ' '));
  } catch {
    return null;
  }
}

// D7's table, byte-for-byte: upper/lower never fail; base64/base64decode/urldecode can, per the
// notes above; urlencode reuses this repo's own byte-for-byte match of Go's url.QueryEscape
// (P12 D7/P9 F16) rather than re-deriving it — a value piped through it is overwhelmingly headed
// for a query string (D7's own stated reason for choosing QueryEscape's space-as-'+' over
// PathEscape's space-as-'%20').
const TRANSFORMS: Readonly<Record<TransformName, (s: string) => string | null>> = {
  base64: base64Encode,
  base64decode: base64Decode,
  upper: (s) => s.toUpperCase(),
  lower: (s) => s.toLowerCase(),
  urlencode: (s) => goQueryEscapeLiteral(s),
  urldecode: urlDecode,
};

/**
 * Applies `pipeline`'s transforms to `value`, left to right — `{{name | base64 | upper}}` first
 * base64-encodes, then upper-cases the result (D7's chaining rule). Returns `null` the moment any
 * step fails (D5: nothing is emitted half-transformed), rather than a partially-applied string.
 * An empty pipeline returns `value` unchanged.
 */
export function applyPipeline(pipeline: readonly TransformName[], value: string): string | null {
  let out = value;
  for (const name of pipeline) {
    const applied = TRANSFORMS[name](out);
    if (applied === null) return null;
    out = applied;
  }
  return out;
}
